azkaban-common/src/main/java/azkaban/scheduler/MissedSchedulesManager.java

```
@@ -0,0 +1,216 @@
+    if (this.missedSchedulesManagerEnabled && threadPoolSize <= 0) {
+      String errorMsg = "MissedSchedulesManager is enabled but thread pool size is not positive: " + threadPoolSize;
+      LOG.error(errorMsg);
+      throw new RuntimeException(errorMsg);
+    }
+  }
+
+  /**
+   * Put timestamps for Missed executions along with email notifications into Task Queue.
+   *
+   * @param missedScheduleTimesInMs, timestamps of missed schedule
+   * @param action, execute flow action
+   * @param backExecutionEnabled, a schedule config from user, the default is false
+   * */
+  public boolean addMissedSchedule(final List<Long> missedScheduleTimesInMs,
+      final ExecuteFlowAction action,
+      final boolean backExecutionEnabled) {
+    String projectId = action.getProjectName();
+    String flowName = action.getFlowName();
+    Project project = projectManager.getProject(projectId);
+    if (project == null) {
+      return false;
+    }
+    List<String> emailRecipients = project.getFlow(flowName).getFailureEmails();
+    try {
+      if (backExecutionEnabled) {
+        LOG.info("Missed schedule task submitted with emailer {} and action {}", emailRecipients, action);
+        Future taskFuture = this.missedSchedulesExecutor.submit(
+            new MissedSchedulesOperationTask(
+                missedScheduleTimesInMs,
+                this.emailer,
+                emailRecipients,
+                action));
+      } else {
+        Future taskFuture = this.missedSchedulesExecutor.submit(
+            new MissedSchedulesOperationTask(
+                missedScheduleTimesInMs,
+                this.emailer,
+                emailRecipients,
+                projectId,
+                flowName));
+      }
+      return true;
+    } catch (RejectedExecutionException e) {
+      LOG.error("Failed to add more missed schedules tasks to the thread pool", e);
```

Maybe rethrow the exception ?

---

Same as above.